libsolidity/ast/ASTJsonImporter.cpp

```diff
@@ -0,0 +1,1003 @@
+/*
+    This file is part of solidity.
+
+    solidity is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    solidity is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author julius <djudju@protonmail.com>
+ * @date 2017
+ * Converts an AST from json format to an ASTNode
+ */
+
+#include <libsolidity/ast/ASTJsonImporter.h>
+#include <libsolidity/parsing/Scanner.h>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string.hpp>
+#include <libsolidity/parsing/Token.h>
+
+
+using namespace std;
+
+namespace
+{
+template<class T>
+dev::solidity::ASTPointer<T> castPointer(dev::solidity::ASTPointer<dev::solidity::ASTNode> _ast)
+{
+	dev::solidity::ASTPointer<T> ret = dynamic_pointer_cast<T>(_ast);
+	if (!ret)
+		solAssert(false, "");
+	else
+		return ret;
+}
+}
+
+namespace dev
+{
+namespace solidity
+{
+
+ASTJsonImporter::ASTJsonImporter(map<string, Json::Value const*> _sourceList )
+        : m_sourceList(_sourceList)
+{
+}
+
+map<string, ASTPointer<SourceUnit>> ASTJsonImporter::jsonToSourceUnit()
+{
+	for (auto const& srcPair: m_sourceList)
+	{
+//		solAssert(!m_json->isNull(),"");
+		solAssert(!srcPair.second->isNull(), "");
+		m_currentSource = srcPair.first;
+		m_sourceUnits[srcPair.first] =  createSourceUnit(*srcPair.second);
+	}
+	return m_sourceUnits;
+}
+
+SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _node)
+{
+	solAssert(!_node["src"].isNull(), "JsonValue should not be an ASTNode");
+        string srcString = _node["src"].asCString();
+        vector<string> pos;
+        boost::algorithm::split(pos, srcString, boost::is_any_of(":"));
+        int start = stoi(pos[0]);
+	int end = start + stoi(pos[1]);
+	shared_ptr<string const> source = make_shared<string const>(m_currentSource);
```

Comment:
It does not have to be the current source, but rather it is an index into the source index list.